### PR TITLE
Generic credential

### PR DESCRIPTION
**Pull request by Wareek**
Replacement of credential management with vault service for username and
password storage

### CODE REVIEW

**Comment by alberto-brigandi** [comment]
Add to whitelist



Alberto Brigandì

Concept Reply
Via Cardinal Massaia, 83
10147 - Torino - ITALY
phone: +39 011 29100
a.brigandi@reply.it<mailto:a.brigandi@reply.it>
www.reply.it<http://www.reply.it> <http://>

[Concept Reply]
________________________________

File: src/main/java/it/reply/orchestrator/dto/vault/VaultTokenExtended.java [hunk 1]
```diff
@@ -0,0 +1,30 @@
...
+ */
+
+package it.reply.orchestrator.dto.vault;
+
+import lombok.Getter;
+
+import org.springframework.vault.support.VaultToken;
+
+@Getter
+public class VaultTokenExtended extends VaultToken {
```
Comment (alberto-brigandi) [inline src/main/java/it/reply/orchestrator/dto/vault/VaultTokenExtended.java]: Do we need this class? Is it used somewhere?

File: src/test/java/it/reply/orchestrator/service/deployment/providers/factory/MarathonClientFactoryTest.java [hunk 2]
```diff
@@ -39,21 +49,39 @@
...
         .cpEndpoint("http://example.com")
         .cpComputeServiceId(UUID.randomUUID().toString())
         .iaasType(IaaSType.MARATHON)
         .build();
+
+    String serviceId = cloudProviderEndpoint.getCpComputeServiceId();
+
+    Mockito
+        .when(credProvServ.credentialProvider(serviceId, "token", GenericServiceCredential.class))
+        .thenReturn(new GenericServiceCredentialWithTenant("username", "password", "tenant"));
```
Comment (alberto-brigandi) [inline src/test/java/it/reply/orchestrator/service/deployment/providers/factory/MarathonClientFactoryTest.java]: `.thenReturn(new GenericServiceCredential("username", "password"));`

File: src/main/java/it/reply/orchestrator/service/deployment/providers/factory/ImClientFactory.java [hunk 3]
```diff
@@ -73,67 +79,113 @@ protected OpenStackCredentials getOpenStackAuthHeader(CloudProviderEndpoint clou
...
+            .map(OpenStackCredentials::buildCredentials)
+            .orElseGet(OpenStackCredentials::buildCredentials)
+            .withTenant(imCred.getTenant())
+            .withUsername(imCred.getUsername())
+            .withPassword(imCred.getPassword())
+            .withHost(endpoint);
+        cloudProviderEndpoint
+            .getRegion()
+            .ifPresent(cred::withServiceRegion);
+        if ("v2".equals(matcher.group(2))) {
```
Comment (alberto-brigandi) [inline src/main/java/it/reply/orchestrator/service/deployment/providers/factory/ImClientFactory.java]: https://imdocs.readthedocs.io/en/latest/client.html#auth-file 
`2.0_password or 3.x_password`

File: src/main/java/it/reply/orchestrator/service/deployment/providers/factory/ImClientFactory.java [hunk 4]
```diff
@@ -50,13 +52,17 @@
...
 import org.checkerframework.checker.nullness.qual.Nullable;
+import org.springframework.beans.factory.annotation.Autowired;
 import org.springframework.stereotype.Service;
 
 @Slf4j
 @Service
 @AllArgsConstructor
 public class ImClientFactory {
 
+  @Autowired
```
Comment (alberto-brigandi) [inline src/main/java/it/reply/orchestrator/service/deployment/providers/factory/ImClientFactory.java]: put this after the static fields and remove the unneeded autowired

**Comment by maricaantonacci** [comment]
Hi @alberto-brigandi and @Wareek 
I've started to test the changes included in this PR and I had to modify the code as follows in order to make it work (only Openstack and AWS have been tested so far):
1. pass the Vault role (read from config) when calling the login API;
2. add the no args constructor for the classes GenericServiceCredential and GenericServiceCredentialWithTenant.

(1) is needed to avoid the permission denied error since the default policy does not allow to read from the path used to store the service credentials.
(2) is needed to fix the following error:
`org.springframework.http.converter.HttpMessageNotReadableException: JSON parse error: Can not construct instance of it.reply.orchestrator.dto.security.GenericServiceCredentialWithTenant: no suitable constructor found, can not deserialize from Object value (missing default constructor or creator, or perhaps need to add/enable type information?); nested exception is com.fasterxml.jackson.databind.JsonMappingException: Can not construct instance of it.reply.orchestrator.dto.security.GenericServiceCredentialWithTenant: no suitable constructor found, can not deserialize from Object value (missing default constructor or creator, or perhaps need to add/enable type information?)
 at [Source: java.io.PushbackInputStream@7038e49c; line: 1, column: 119] (through reference chain: org.springframework.vault.support.VaultResponseSupport["data"])
	at org.springframework.http.converter.json.AbstractJackson2HttpMessageConverter.readJavaType(AbstractJackson2HttpMessageConverter.java:240)
	at org.springframework.http.converter.json.AbstractJackson2HttpMessageConverter.read(AbstractJackson2HttpMessageConverter.java:225)
	at org.springframework.web.client.HttpMessageConverterExtractor.extractData(HttpMessageConverterExtractor.java:96)
	at org.springframework.web.client.RestTemplate$ResponseEntityResponseExtractor.extractData(RestTemplate.java:934)
	at org.springframework.web.client.RestTemplate$ResponseEntityResponseExtractor.extractData(RestTemplate.java:918)
	at org.springframework.web.client.RestTemplate.doExecute(RestTemplate.java:664)
	at org.springframework.web.client.RestTemplate.execute(RestTemplate.java:622)
	at org.springframework.web.client.RestTemplate.exchange(RestTemplate.java:568)
	at org.springframework.vault.core.VaultTemplate.read(VaultTemplate.java:238)
	at it.reply.orchestrator.service.VaultServiceImpl.readSecret(VaultServiceImpl.java:104)
	at it.reply.orchestrator.service.VaultServiceImpl.readSecret(VaultServiceImpl.java:109)
	at it.reply.orchestrator.service.deployment.providers.CredentialProviderServiceImpl.credentialProvider(CredentialProviderServiceImpl.java:69)
	at it.reply.orchestrator.service.deployment.providers.factory.ImClientFactory.getOpenStackAuthHeader(ImClientFactory.java:107)`

**Pull request merged**